roles/kafka_broker/tasks/main.yml

```diff
@@ -231,6 +231,35 @@
   tags:
     - configuration
 
+- name: get uuid
+  shell: "{{ binary_base_path }}/bin/kafka-storage random-uuid"
+  environment:
+    KAFKA_OPTS: "-Xlog:all=error -XX:+IgnoreUnrecognizedVMOptions"
+  register: uuid_key
+  run_once: true
+  when: kraft_enabled|bool and inventory_hostname in groups.kafka_controller
+
+- name: extract uuid from meta.properties
+  slurp:
+    src: "{{ kafka_broker_final_properties['log.dirs'].split(',')[0] }}/meta.properties"
+  delegate_to: "{{ groups.kafka_controller[0] }}"
+  register: uuid_broker
+  when: kraft_enabled|bool and inventory_hostname in groups.kafka_broker
+
+- set_fact:
+    uuid_key_broker: "{{ uuid_broker['content'] | b64decode}} "
+  when: kraft_enabled|bool and inventory_hostname in groups.kafka_broker
+
+- name: get CLUSTER_ID
+  set_fact:
+    clusterid: "{{ uuid_key.stdout if inventory_hostname in groups.kafka_controller else uuid_key_broker.partition('cluster.id=')[2].partition('\n')[0] }}"
+  when: kraft_enabled|bool
+
+- name: format meta directory
+  shell: "{{ binary_base_path }}/bin/kafka-storage  format -t {{ clusterid }} -c {{ config_base_path }}/etc/kafka/server.properties --ignore-formatted"
+  register: format_meta
+  when: kraft_enabled|bool
+
 # Zookeeper shell cannot read server.properties if secrets protection is enabled. Keeping a separate file.
 - name: Create Zookeeper TLS Client Config
```

Comment:
You can separate out these tasks in another play and have this condition at a single place.